src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/SpanNDHelpers.cs

```diff
@@ -0,0 +1,508 @@
+                    //     +-------+
+                    //     | A | B |
+                    //     +-------+
+                    //          --->
+                    //     +-------+
+                    //     | B | A |
+                    //     +-------+
+                    tempFirst = Vector128.Shuffle(tempFirst, Vector128.Create(1, 0));
+                    tempLast = Vector128.Shuffle(tempLast, Vector128.Create(1, 0));
+
+                    // Store the reversed vectors
+                    tempLast.StoreUnsafe(ref buf, (nuint)offset);
+                    tempFirst.StoreUnsafe(ref buf, (nuint)lastOffset);
+
+                    offset += Vector128<long>.Count;
+                    lastOffset -= Vector128<long>.Count;
+                } while (lastOffset >= offset);
+
+                remainder = lastOffset + Vector128<long>.Count - offset;
+            }
+
+            // Store any remaining values one-by-one
+            if (remainder > 1)
+            {
+                ReverseInner(ref Unsafe.Add(ref buf, offset), (nuint)remainder);
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static unsafe void Reverse<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 1);
+
+            if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            {
+                if (sizeof(T) == sizeof(byte))
+                {
+                    Reverse(ref Unsafe.As<T, byte>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(char))
+                {
+                    Reverse(ref Unsafe.As<T, char>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(int))
+                {
+                    Reverse(ref Unsafe.As<T, int>(ref elements), length);
+                    return;
+                }
+                else if (sizeof(T) == sizeof(long))
+                {
+                    Reverse(ref Unsafe.As<T, long>(ref elements), length);
+                    return;
+                }
+            }
+
+            ReverseInner(ref elements, length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void ReverseInner<T>(ref T elements, nuint length)
+        {
+            Debug.Assert(length > 1);
+
+            ref T first = ref elements;
+            ref T last = ref Unsafe.Subtract(ref Unsafe.Add(ref first, length), 1);
+            do
+            {
+                T temp = first;
+                first = last;
+                last = temp;
+                first = ref Unsafe.Add(ref first, 1);
+                last = ref Unsafe.Subtract(ref last, 1);
+            } while (Unsafe.IsAddressLessThan(ref first, ref last));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static nint CalculateTotalLength(ref nint[] lengths)
```

Comment:
It's not clear why `lengths` needs to be a ref and why this can't removed in favor of only having the one that takes a `ROSpan<nint>` instead